### PR TITLE
Changes to create-native-map

### PR DESCRIPTION
**Pull request by steffen-kiess**
- Use CompareOptions.OrdinalIgnoreCase instead of
  CompareOptions.Ordinal | CompareOptions.IgnoreCase (which is no longer
  supported by newer mono versions)
  This is needed to make the code work on current mono master.

- Make sure enum hex values are lowercase on newer mono versions
  This is needed to make the code produce the lowercase enum values (like they are in the current mono git) on current mono master.

These changes are needed for a the second part of the posix socket pull request:
- Do not create conversion calls for inherited fields
- Allow [Map] for internal enums

### CODE REVIEW

**Comment by steffen-kiess** [comment]
See https://github.com/mono/mono/pull/2006 for the pull request this is needed for

**Comment by jonpryor** [comment]
Hard to believe as it may be, there are unit tests. ;-)

There's a [`make check`](https://github.com/mono/mono-tools/blob/master/create-native-map/src/Makefile.am#L47-L50) target which runs `create-native-map` against a [test input file](https://github.com/mono/mono-tools/blob/master/create-native-map/src/TestMap.cs) and compares it to expected output, e.g. [test.c.ref](https://github.com/mono/mono-tools/blob/master/create-native-map/src/test.c.ref).

Please update the expected output files to match the new output.

**Comment by steffen-kiess** [comment]
> Hard to believe as it may be, there are unit tests. ;-)

Ah, I really didn't expect that :-)

I've fixed the tests (the enums in TestMap.cs now have to be public to create the same code) and added tests for the new functionality.

**Comment by jonpryor** [comment]
> Do not create conversion calls for inherited fields

I'm not sure this part makes sense, e.g. from the generated test code:

	[Map ("struct parent")]
	[StructLayout (LayoutKind.Sequential)]
	public class Parent {
		public int i;
	}

	[Map ("struct child")]
	[StructLayout (LayoutKind.Sequential)]
	public class Child : Parent {
		public int j;
	}

`Child` inherits `Parent`, okay...

	struct MakeMap_Test_Parent {
		int i;
	};
	...
	struct MakeMap_Test_Child {
		int i;
		int j;
	};

The `Child` gets all the members of `Parent`; okay...

	int
	MakeMap_Test_FromChild (struct MakeMap_Test_Child *from, struct child *to)
	{
		_cnm_return_val_if_overflow (int, from->j, -1);
	
		memset (to, 0, sizeof(*to));
	
		to->j = from->j;
	
		return 0;
	}

Copying a `Child` instance from a `struct child` *ignores the inherited members*; `struct child::i` isn't copied!

This part makes no sense to me.

**Comment by steffen-kiess** [comment]
> Copying a `Child` instance from a `struct child` ignores the inherited members; `struct child::i` isn't copied!
> 
> This part makes no sense to me.

The issue is that `struct child` does not necessarily have a member `i`. `struct parent` has a member `i`, and POSIX never seems to use the same member name for different structs (all member names have a prefix indicating the struct the member is in) so `struct child` probably does not have such a member.

This is used for `struct sockaddr` and e.g. `struct sockaddr_in`. `struct sockaddr` (and therefore the `Sockaddr` class in C#) has the member `sa_family`. `struct sockaddr_in` however does not have a `sa_family` member (instead it has a `sin_family` member at the same offset), therefore the conversion code must not try to copy `sa_family`. The `sa_family` field itself is currently handled manually in the C helper code.

So basically the idea is that it is up the caller to handle the values of the parent classes because the code generated by create-native-map doesn't know enough and because for the only case where this is currently used (`struct sockaddr`) there is only one field to copy anyway.

**Comment by jonpryor** [comment]
> So basically the idea is that it is up the caller to handle the values of the parent classes because the code generated by create-native-map doesn't know enough...

I see. That makes sense.

So how do we make this understandable? The implemented behavior isn't exactly "obvious," in so far as I didn't understand the behavior. ;-)

I can think of a few approaches:

1. Documentation. We actually have [documentation](https://github.com/mono/mono-tools/blob/master/create-native-map/man/create-native-map.1)...which is apparently available [on ubuntu.com](http://manpages.ubuntu.com/manpages/gutsy/man1/create-native-map.1.html) (?!). We could document this behavior.

    I forgot that we even *had* docs (lol!). The downside to this is the need to write man-page syntax.

2. Generate a warning. This sucks, because the warning would just tell you what's going on, and nobody ever reads warnings, *ever*, but hey (!), it's a "fix."

3. Alter `MapAttribute` so that the generator behavior is explicitly listed, e.g. via a `MapAttribute.CopyInheritedMembers` property:

        partial class MapAttribute {
            public bool CopyInheritedMembers {get; set;} = true;
        }

    I'm not sure which is the better default here, `true` or `false`, but at least it allows readers of the usage site to know what will happen:

		[Map ("struct sockaddr_in6", CopyInheritedMembers=false)]
		[StructLayout (LayoutKind.Sequential)]
		public partial class SockaddrIn6 : Sockaddr, IEquatable<SockaddrIn6> {
			// ...
		}

Given that there's no current need to implement copying of inherited members, (3) is likely premature design. (2) sucks.

Would you be willing to try (1)?

**Comment by steffen-kiess** [comment]
Ok, I've added a paragraph in the manpage.

I also could do (3) but it's not unlikely that no one will ever need it, so it's probably better to wait.
